File: app/src/main/java/com/rafaeltmbr/stopwatch/domain/use_cases/impl/StartStopwatchUseCaseImpl.kt
package com.rafaeltmbr.stopwatch.domain.use_cases.impl

import com.rafaeltmbr.stopwatch.domain.entities.Status
import com.rafaeltmbr.stopwatch.domain.entities.StopwatchState
import com.rafaeltmbr.stopwatch.domain.services.TimerService
import com.rafaeltmbr.stopwatch.domain.stores.MutableStateStore
import com.rafaeltmbr.stopwatch.domain.use_cases.StartStopwatchUseCase
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class StartStopwatchUseCaseImpl(
    private val store: MutableStateStore<StopwatchState>,
    private val timerService: TimerService,
    coroutineScope: CoroutineScope = CoroutineScope(Dispatchers.Default)
) : StartStopwatchUseCase {
    init {
        coroutineScope.launch {
            timerService.state.collect { timerState ->
                store.update { it.copy(milliseconds = timerState.milliseconds) }
            }
        }
    }

    override suspend fun execute() {
        if (store.state.value.status == Status.RUNNING) return

        timerService.start()
        store.update { it.copy(status = Status.RUNNING) }
    }
}

File: ios/Stopwatch/domain/use_cases/impl/StartStopwatchUseCaseImpl.swift
class StartStopwatchUseCaseImpl<MSS, TS>: StartStopwatchUseCase
where MSS: MutableStateStore, MSS.State == StopwatchState, TS: TimerService {
    private let store: MSS
    private let timer: TS
    
    init(_ store: MSS, _ timer: TS) {
        self.store = store
        self.timer = timer
    }
    
    func execute() async {
        guard !timer.state.isRunning else { return }
        
        await store.update {
            StopwatchState(
                status: .running,
                milliseconds: $0.milliseconds,
                completedLaps: $0.completedLaps,
                completedLapsMilliseconds: $0.completedLapsMilliseconds
            )
        }
        timer.start()
    }
}
